test/parallel/test-async-wrap-init-hook.js

```diff
@@ -0,0 +1,33 @@
+var common = require('../common');
+var assert = require('assert');
+
+var asyncWrap = process.binding('async_wrap');
+
+var asyncHooksObject = {};
+var kCallInitHook = 0;
+asyncWrap.setupHooks(asyncHooksObject, init, before, after);
+asyncHooksObject[kCallInitHook] = 1;
+
+var order = [];
+
+function init() {
+  order.push('init ' + this.constructor.name);
+}
+
+function before() {
+  order.push('before ' + this.constructor.name);
+}
+
+function after() {
+  order.push('after ' + this.constructor.name);
+}
+
+setTimeout(function () {
+  order.push('callback');
+});
```

Comment:
Just for a future note, this may be spotty because there's incomplete timer support. Because only a single `TimerWrap` is instantiated for many timers. Still need to implement the JS side for proper full support.